DOC Update "Parallelism, resource management, and configuration" section (#24997)
scikit-learn/scikit-learn: doc/computing/parallelism.rst

Parallelism
-----------

Some scikit-learn estimators and utilities parallelize costly operations
using multiple CPU cores.

> This can be either done:

Reviewer: Let's not make implicitly suggest that the user can always choose:

```suggestion
Depending on the type of estimator and sometimes the values of the
constructor parameters, this is either done:
```

- with higher-level parallelism via `joblib <https://joblib.readthedocs.io/en/latest/>`_.
  In this case, the number of threads or processes can be controlled with the
  ``n_jobs`` parameter.
- with lower-level parallelism via OpenMP, used in C or Cython code.
  In this case, parallelism is always done using threads and specifying

>   ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally

Author: That's it. It is just that `n_jobs` has currently be chosen to not be propagated to indicate the number of threads to use for OpenMP.

Personally, I think it makes sense to have `n_jobs` specify the number of threads in this case (and thus have the documentation and implementation of related user-facing API changed accordingly); but this is to be discussed among maintainers.

This document from @thomasjpfan gives some perspective on the semantic: https://thomasjpfan.github.io/parallelism-python-libraries-design/

>   ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally

Reviewer 2: If setting `n_jobs` does nothing, then why does the parameter exist and why does it not raise a deprecation warning? Do you have a link to an estimator as an example? Maybe that helps to understand what is happening.

>   ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally

Reviewer 3: `n_jobs` does have an effect on OpenMP parallelism. For example in `RandomSearchCV`, `joblib` will reduce the number of OpenMP threads to prevent oversubscription:

XREF: https://github.com/joblib/joblib/blob/796e7aa5c6f5a19045f06ad65415e170ae6f8e9a/joblib/parallel.py#L189

>   ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally

Author: Thanks for mentioning it.

So `n_jobs` only has an impact on the number of threads spawn by OpenMP if joblib is used? I cannot remember of a scenario where `n_jobs` specify behavior for OpenMP directly without joblib (or any of joblib's backends).

>   ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally

Reviewer 2: Thanks for the details. Now it makes sense. I had understood your original comment as "the parameter exists but is never used anymore". However that isn't what happens. It is used sometimes, but there are now cases where it does nothing. As well as cases where it changes the number of threads used by OpenMP.

Sounds like overall it is quite tricky to know when `n_jobs` has an effect and if it does, what exactly it will do. Fun for the docs :D

>   ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally

Reviewer: > So n_jobs only has an impact on the number of threads spawn by OpenMP if joblib is used? I cannot remember of a scenario where n_jobs specify behavior for OpenMP directly without joblib (or any of joblib's backends).

`n_jobs` set to control the amount of **outer** parallelism (with `loky`), for instance in `RandomSearchCV` has an indirect impact on the amount of Cython-OpenMP threading parallelism used in the inner tuned `HistGradientBoostingClassifier` or OpenBLAS threading parallelism of any estimators that relies on numpy/scipy BLAS operations for the heavy lifting inside the loky workers.

>   ``n_jobs`` *has no effect*. Implementations relying on this parallelism are generally

Reviewer: Instead of stating:

"In this case, parallelism is always done using threads and specifying ``n_jobs`` *has no effect*"

one could add a new sentence as a new paragraph after the end of the 2 bullet point such as:

> The `n_jobs` parameters of estimators always controls the amount of parallelism managed by joblib (processes or threads depending on the joblib backend). The thread-level parallelism managed by OpenMP in scikit-learn's own Cython code or by BLAS/LAPACK libraries used by numpy and scipy operations used in scikit-learn is always controlled by environment variables or `threadpoolctl` as explained below. Note that some estimators can leverage all 3 kinds of parallelism at different points of their training and prediction methods.

>   more performant by several orders of magnitude.

Author: ```suggestion
  more performant by up to two orders of magnitude.
```

>   more performant by several orders of magnitude.

Reviewer 2: More performant than what? A group of monkeys with calculators? :D

Author: Perfect timing for using this emoji: :see_no_evil: 

Is this clearer? :upside_down_face: 
```suggestion
  more performant than joblib-based implementations by several orders of magnitude.
```

>   more performant than joblib-based implementations by up to two orders of magnitude.

Reviewer: I would remove this last sentence because those "joblib-based implementations" are no longer available to the scikit-learn user. I think this can cause confusion for users who might wondering where they are and how they can be enabled.

The user guide should only refer to the current state of the library, not about the evolution between releases. The changelog and blog posts are better for that.

Also the main reason why the new Cython implementations are orders of magnitude faster than previous numpy-based implementations is primarily because they rely on fine-grained chunking to avoid RAM transfers temporary arrays and not necessarily because of joblib-managed threads vs OpenMP-managed threads. Those (important) implementation details are outside of the scope of this paragraph.

Reviewer: Instead of stating:

"In this case, parallelism is always done using threads and specifying ``n_jobs`` *has no effect*"

one could add a new sentence as a new paragraph after the end of the 2 bullet point such as:

> The `n_jobs` parameters of estimators always controls the amount of parallelism managed by joblib (processes or threads depending on the joblib backend). The thread-level parallelism managed by OpenMP in scikit-learn's own Cython code or by BLAS/LAPACK libraries used by numpy and scipy operations used in scikit-learn is always controlled by environment variables or `threadpoolctl` as explained below. Note that some estimators can leverage all 3 kinds of parallelism at different points of their training and prediction methods.

Author: > I would remove this last sentence because those "joblib-based implementations" are no longer available to the scikit-learn user. I think this can cause confusion for users who might wondering where they are and how they can be enabled.

There's an option in the scikit-learn configuration to disable `PairwiseDistancesReduction` if needed.

> Also the main reason why the new Cython implementations are orders of magnitude faster than previous numpy-based implementations is primarily because they rely on fine-grained chunking to avoid RAM transfers temporary arrays and not necessarily because of joblib-managed threads vs OpenMP-managed threads. Those (important) implementation details are outside of the scope of this paragraph.

It is true that this is out of the scope of this section.

estimators or functions in parallel (see oversubscription below).

Lower-level parallelism with OpenMP
...................................

OpenMP is used to parallelize code written in Cython or C, relying on
multi-threading exclusively. By default, the implementations using OpenMP
will use as many threads as possible, i.e. as many threads as logical cores.

You can control the exact number of threads that are used either:

 - via the ``OMP_NUM_THREADS`` environment variable, e.g. for instance when:
   running a python script:

   .. prompt:: bash $

        OMP_NUM_THREADS=4 python my_script.py

 - or via `threadpoolctl` as explained by `this piece of documentation
   <https://github.com/joblib/threadpoolctl/#setting-the-maximum-size-of-thread-pools>`_.

Parallel NumPy and SciPy routines from numerical libraries
..........................................................

scikit-learn relies heavily on NumPy and SciPy, which internally call
multi-threaded linear algebra routines of BLAS implemented in libraries
such as MKL, OpenBLAS or BLIS.

You can control the exact number of threads used by BLAS for each library

> using environment variables, namely:
>   - ``MKL_NUM_THREADS`` sets the number of thread MKL uses,

Author: ```suggestion
using environment variables, namely:

  - ``MKL_NUM_THREADS`` sets the number of thread MKL uses,
```

it can be highly detrimental to performance to run multiple copies of some
estimators or functions in parallel (see oversubscription below).

Lower-level parallelism with OpenMP
...................................

OpenMP is used to parallelize code written in Cython or C, relying on
multi-threading exclusively. By default, the implementations using OpenMP
will use as many threads as possible, i.e. as many threads as logical cores.

You can control the exact number of threads that are used either:

>  - via the ``OMP_NUM_THREADS`` environment variable, e.g. for instance when:

Reviewer 2: "e.g. for instance" == "for example for instance" =>

```suggestion
 - via the ``OMP_NUM_THREADS`` environment variable, for instance when:
```

Author: Ah yes, duplication due to style switch!

   running a python script:

   .. prompt:: bash $

        OMP_NUM_THREADS=4 python my_script.py

 - or via `threadpoolctl` as explained by `this piece of documentation
   <https://github.com/joblib/threadpoolctl/#setting-the-maximum-size-of-thread-pools>`_.

Parallel NumPy and SciPy routines from numerical libraries
..........................................................

scikit-learn relies heavily on NumPy and SciPy, which internally call

> multi-threaded linear algebra routines of BLAS implemented in libraries

Reviewer: ```suggestion
multi-threaded linear algebra routines (BLAS & LAPACK) implemented in libraries
```

such as MKL, OpenBLAS or BLIS.

You can control the exact number of threads used by BLAS for each library
using environment variables, namely:

  - ``MKL_NUM_THREADS`` sets the number of thread MKL uses,
  - ``OPENBLAS_NUM_THREADS`` sets the number of threads OpenBLAS uses

>   - ``BLIS_NUM_THREADS`` sets the number of threads BLIS uses

Reviewer: ```suggestion
  - ``BLIS_NUM_THREADS`` sets the number of threads BLIS uses

Note that BLAS & LAPACK implementations can also be impacted by
`OMP_NUM_THREADS`. To check whether this is the case in your environment,
you can inspect how the number of threads effectively used by those libraries
is affected when running the the following command in a bash or zsh terminal
for different values of `OMP_NUM_THREADS`::

    OMP_NUM_THREADS=2 python -m threadpoolctl -i numpy scipy
```

Author: I just have edited this suggestion to add the `.. prompt:: bash $` directive.

  only use ``<LIB>_NUM_THREADS``. Joblib exposes a context manager for
  finer control over the number of threads in its workers (see joblib docs
  linked below).
- `threadpoolctl` internally manages the numbers of threads used by OpenMP
  and BLAS implementations for scikit-learn implementations.

You will find additional details about joblib mitigation of oversubscription
in `joblib documentation
<https://joblib.readthedocs.io/en/latest/parallel.html#avoiding-over-subscription-of-cpu-resources>`_.

You will find additional details about parallelism in numerical python libraries
in `this document from Thomas J. Fan <https://thomasjpfan.github.io/parallelism-python-libraries-design/>`_.

Configuration switches
-----------------------

Python runtime
..............

> :func:`sklearn.set_config` controls the following behaviors.

Author: Here, I think it makes sense to explain how `sklearn.{set_config,get_config,config_context}` can be used and point to their documentation rather than duplicating parameters' documentation.

What do you think?

Reviewer 4: Make sense. We might only add which behaviours can be controlled: finiteness, memory, type of output, etc.

Author: ede91d210b1fd7bfd0d388069297e3f62c6ed2ea gave some precisions.

  only use ``<LIB>_NUM_THREADS``. Joblib exposes a context manager for
  finer control over the number of threads in its workers (see joblib docs
  linked below).

> - `threadpoolctl` internally manages the numbers of threads used by OpenMP
>   and BLAS implementations for scikit-learn implementations.

Reviewer: The original item was about the lack of builtin mechanism to avoid oversubscription when using the threading backend of joblib.

```suggestion
- When joblib is configured to use the ``threading`` backend, there is no
  mechanism to avoid oversubscriptions when calling into parallel native
  libraries in the joblib-managed threads.
- All scikit-learn estimators that explicitly rely on OpenMP in their Cython code
  always use `threadpoolctl` internally to automatically adapt the numbers of
  threads used by OpenMP and potentially nested BLAS calls so as to avoid
  oversubscription.
```
